### ISSUE TITLE
torch.compile does not work for layer drop

### ISSUE THREAD

**Issue by wconstab** [pytorch/pytorch#95425]
### 🐛 Describe the bug

In the huggingface model `MBartForConditionalGeneration`, which is part of the benchmark suite running on PT2, layer drop is used.  I've extracted a minimal repro of layer drop inspired by `MBartForConditionalGeneration` and pasted below.

In both cases, dynamo will evaluate the random number during tracing and choose whether to trace a layer or skip it.  The random number is evaluated during tracing and not saved as part of the graph, so there is no chance to guard or catch this at runtime.  Instead, each execution of the model will use the frozen trace from the first set of dice rolls.

One fix for this could be to graph-break on random.uniform calls.  Unfortunately doing this for the MBart model causes our CI to OOM due to the excessive number of graph breaks.

```
import random
import torch
from torch import nn
import torch._dynamo.testing

class MyModule(torch.nn.Module):
    def __init__(self, a, b):
        super().__init__()
        self.net = nn.Sequential(
            nn.Linear(a, b),
            nn.ReLU(),
        )

    def forward(self, x):
        return self.net(x)


class ToyModel(nn.Module):
    def __init__(self):
        super().__init__()
        self.pre = nn.Sequential(
            *[nn.Linear(10, 10000), nn.ReLU()]
            + [nn.Linear(10000, 10000), nn.ReLU()]
            + [MyModule(10000, 10000)]
            + [MyModule(10000, 1000)]
        )

        self.layers = [
            MyModule(1000, 1000),
            MyModule(1000, 1000),
            MyModule(1000, 1000),
            MyModule(1000, 1000),
            MyModule(1000, 1000),
            MyModule(1000, 1000),
        ]
       
        self.tail = nn.Sequential(
            *[nn.Linear(1000, 5)]
        )

        self.layerdrop = 0.5

    def forward(self, x):
        hidden_state = self.pre(x)
        for layer in self.layers:

            dropout_probability = random.uniform(0, 1)

            if dropout_probability < self.layerdrop:
                continue

            hidden_state = layer(hidden_state) 
        
        hidden_state = self.tail(hidden_state)
        return hidden_state


m = ToyModel()
inputs = (torch.randn(20, 10),)
cnt = torch._dynamo.testing.CompileCounter()
opt_m = torch._dynamo.optimize(cnt)(m)
out = opt_m(*inputs)
assert cnt.frame_count == 1, "captured a full graph on the first try"

out = opt_m(*inputs)
assert cnt.frame_count == 2, "Expected a recompile on each iteration, since there are 2^n combinations of layerdrop"

```

### Versions

torch@ bc438af6fed4fe1fd0ed80e6d5f5ea17c3ca30bb

cc @ezyang @soumith @msaroufim @ngimel @bdhirsh @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

**Comment by wconstab** [comment]
btw, this issue got fixed accidentally by https://github.com/pytorch/pytorch/pull/92125 since it caused more things to get inlined and the inliner properly broke out when it saw random.

however, the current state of 92125 'fixes' this behavior change and keeps MBart running as it is on master.  One option to fix this issue is to land 92125 and take the special case out, which will graph break and make the model run correctly on a local machine with enough gpu ram.  We might not be able to run the model in CI though.

**Comment by yanboliang** [comment]
@wconstab 
For python builtin random calls like ```random.uniform```, we don't put it in FX graph. We made this decision based on these functions are python builtin, not torch random functions which should be part of FX graph.
https://github.com/pytorch/pytorch/blob/1e3abda31a51f6efb0c8e886a7eb4e728b80ac64/torch/_dynamo/variables/user_defined.py#L278-L293
We can revisit this decision, it's not hard to put these functions in FX graph if we think it makes more sense.

**Comment by wconstab** [comment]
> We can revisit this decision, it's not hard to put these functions in FX graph if we think it makes more sense.

i don't think you can ever put py rand into the graph.  Dynamo can put it in the fx graph it creates, but then AOT is gonna trace through it and burn in a constant, which will again break layer-drop.

The only correct thing to do here is graph break and keep the py-rand call running every iter in eager and feeding the random # into the graph. (or, we need to put real support for py random into the graph _and_ our backends, avoiding burning a constant in).

**Comment by yanboliang** [comment]
> i don't think you can ever put py rand into the graph. Dynamo can put it in the fx graph it creates, but then AOT is gonna trace through it and burn in a constant, which will again break layer-drop.

I see, so it seems the blocker is AOT doesn't support trace through python rand functions. Does AOT has plan to support this? cc @Chillee @bdhirsh 

> The only correct thing to do here is graph break and keep the py-rand call running every iter in eager and feeding the random # into the graph.

Yes, I think that definitely works. I don't have strong opinion here, but I'm a bit confused about how to distinguish python rand vs torch rand from users' perspective. It seems we already put torch rand in FX graph, do we need do the same thing for python rand as well?

**Comment by wconstab** [comment]
so, the reason we can put torch rand in the graph is that we define it in our core ATen IR and expect our backends to understand it.

the contract for dynamo-AOT is simple: dynamo can either put torch-operators in the graph it gives to AOT, knowing AOT will either preserve or safely decompose them.  or it can put arbitrary functions in the graph if they can be safely traced by AOT, meaning that AOT will reduce them to functionally equivalent torch operations.

python rand doesn't meet either of these criteria- it needs to rerun on every execution, and it has no lowering to a torch/aten operator.

The only hope besides graph break is to define some semantic conversion of python rand into torch rand.  i'm not sure how far apart the two are or if this is at all feasible.

**Comment by yanboliang** [comment]
> python rand doesn't meet either of these criteria- it needs to rerun on every execution, and it has no lowering to a torch/aten operator.

That makes sense, thanks for your detailed explanation!
But I found another way that can satisfy your requirement and no graph break. The current implementation is based on the assumption that every iteration the python rand should return the same result. Look at the following lines, we saved the python rand initial state:
https://github.com/pytorch/pytorch/blob/1e3abda31a51f6efb0c8e886a7eb4e728b80ac64/torch/_dynamo/variables/user_defined.py#L286-L287

I think this maybe not correct, we should not save/use the same initial state for every iteration. The python rand state is updated after the first call, so the second call should use the updated state rather than the saved initial state before the first call.
But we don't put it in FX graph so export mode can't use it, however, the graph break solution can't help export mode as well. If this looks good to you, I can submit a fix.

**Comment by ngimel** [comment]
@yanboliang what is the proposed fix to avoid graph break?

**Comment by yanboliang** [comment]
@ngimel
Current for python rand functions, we save the function calls and initial rand state during compilation. Then for every iteration, we call the rand functions based on the same initial rand state, which is the root cause of this issue. My proposal is:
* Remove the saved initial rand state, for every iteration we use the initial state at runtime, then the rand calls will have different outputs.
* We don't guard on rand calls or initial state right now, this is somehow not correct. Take this issue as example, the rand state of the next iteration should be different from the first one(compilation one), but we don't trigger re-compile due to lack of guard.

**Comment by ngimel** [comment]
If you guard on rand state, you'd need to recompile every time, so I don't see a way around graph break.

**Comment by yanboliang** [comment]
> If you guard on rand state, you'd need to recompile every time, so I don't see a way around graph break.

Oh yes, I think we should not guard on it, but every iteration should generate different random value. Then it would be the same behavior as fall back, but no graph break.

**Comment by yanboliang** [comment]
Actually to make my proposal more clear, let's use the following example:
```
import torch
import random

@torch.compile(backend="eager")
def fn(x):
    return (x + 1) * random.uniform(0, 1)

x = torch.ones(2)
for _ in range(3):
    print(fn(x))
```
Output of current master branch(no re-compilation):
```
tensor([0.7291, 0.7291])
tensor([0.7291, 0.7291])
tensor([0.7291, 0.7291])
```
After applying my fix(no re-compilation):
```
tensor([1.7749, 1.7749])
tensor([1.5697, 1.5697])
tensor([1.1214, 1.1214])
```

The change I did is just remove the function call of setting initial rand state:
```
--- a/torch/_dynamo/output_graph.py
+++ b/torch/_dynamo/output_graph.py
@@ -589,14 +589,14 @@ class OutputGraph(fx.Tracer, Checkpointable[OutputGraphState]):
             rand_fn = disable(_get_gen_rand_values_fn(tx.random_calls))
             self.install_global(rand_fn_name, rand_fn)
             codegen = PyCodegen(tx, root)
-            random_calls_instructions.extend(
-                [
-                    codegen.create_load_global("random", True, add=True),
-                    codegen.create_load_attr("setstate"),
-                    codegen.create_load_const(tx.output.initial_random_state),
-                ]
-                + create_call_function(1, False),
-            )
+            # random_calls_instructions.extend(
+            #     [
+            #         codegen.create_load_global("random", True, add=True),
+            #         codegen.create_load_attr("setstate"),
+            #         codegen.create_load_const(tx.output.initial_random_state),
+            #     ]
+            #     + create_call_function(1, False),
+            # )
             random_calls_instructions.extend(
                 codegen.load_function_name(rand_fn_name, True)
             )
```

**Comment by wconstab** [comment]
I didn't realize how the rand handling worked.  It sounds like the value was not being burned into the graph during tracing as i had assumed.  Instead, dynamo generates bytecode that runs before calling the compiled fx graph and that bytecode generates a sequence of rand calls that are then fed as inputs to the fx graph?

Indeed this sounds like a plausible solution.  I'm not sure why the initial code would have frozen the seed, though.  Can you confirm why that was done (in case this will break some other use case?)

**Comment by yanboliang** [comment]
> I didn't realize how the rand handling worked. It sounds like the value was not being burned into the graph during tracing as i had assumed. Instead, dynamo generates bytecode that runs before calling the compiled fx graph and that bytecode generates a sequence of rand calls that are then fed as inputs to the fx graph?

Yes, this is because we shouldn't burn the value during compilation.

> Indeed this sounds like a plausible solution. I'm not sure why the initial code would have frozen the seed, though. Can you confirm why that was done (in case this will break some other use case?)

I implemented this feature when I just joined our team, I didn't remember there is strong reason that we need to keep the initial state, and actually I think it's not a correct decision now. Let me wrap up a PR and go with CI to see if we break other cases.
